### PR TITLE
Feature/rosboard player

### PR DESCRIPTION
**Pull request by franyol**
# Add a new custom Rosboard Player
**Changelog**

* [New Rosboard client module](packages/studio-base/src/players/rosboardClient.ts) was implemented to interact with the websocket custom fork https://github.com/kiwicampus/rosboard
* [Rosbridge player implementation](packages/studio-base/src/players/RosbridgePlayer.ts) was modified to use the custom rosbridge-websocket client

**Test**
Use the following format to a device websocket connected to rosboard:
`ws://<robot-ip>/rosboard/v1`

The websocket client may not be able to connect to the server when testing it locally, in that case try proxy-ing the websocket through another port also running in localhost.

### CODE REVIEW

File: packages/studio-base/src/players/RosbridgePlayer.ts [hunk 1]
```diff
@@ -546,85 +704,26 @@ export default class RosbridgePlayer implements Player {
...
   public setPublishers(publishers: AdvertiseOptions[]): void {
-    // Since `setPublishers` is rarely called, we can get away with just throwing away the old
-    // Roslib.Topic objects and creating new ones.
-    for (const publisher of this.#topicPublishers.values()) {
-      publisher.unadvertise();
-    }
-    this.#topicPublishers.clear();
-    this.#advertisements = publishers;
-    this.#setupPublishers();
+    return;
```
Comment (charlielito) [inline packages/studio-base/src/players/RosbridgePlayer.ts]: add here a TODO for the future

File: packages/studio-base/src/players/RosbridgePlayer.ts [hunk 2]
```diff
@@ -546,85 +704,26 @@ export default class RosbridgePlayer implements Player {
...
-      if (this.#advertisements.some((opts) => opts.topic === topic)) {
-        // Topic was advertised but the connection is not yet established
-        return;
-      }
-      throw new Error(
-        `Tried to publish on a topic that is not registered as a publisher: ${topic}`,
-      );
-    }
-    publisher.publish(msg);
+    return;
```
Comment (charlielito) [inline packages/studio-base/src/players/RosbridgePlayer.ts]: Add here also a TODO for future implementation

File: packages/studio-base/src/players/RosbridgePlayer.ts [hunk 3]
```diff
@@ -546,85 +704,26 @@ export default class RosbridgePlayer implements Player {
...
-    if (existing) {
-      return await existing;
-    }
-
-    const rosClient = this.#rosClient;
-    const serviceTypePromise = new Promise<string>((resolve, reject) => {
-      rosClient.getServiceType(service, resolve, reject);
-    });
+    // Yep
+    throw new Error("Not connected");
```
Comment (charlielito) [inline packages/studio-base/src/players/RosbridgePlayer.ts]: Add also a TODO here

File: packages/studio-base/src/players/RosbridgePlayer.ts [hunk 4]
```diff
@@ -546,85 +704,26 @@ export default class RosbridgePlayer implements Player {
...
-        request,
-        (response: Record<string, unknown>) => {
-          resolve(response);
-        },
-        (error: Error) => {
-          reject(error);
-        },
-      );
-    });
+    return;
```
Comment (charlielito) [inline packages/studio-base/src/players/RosbridgePlayer.ts]: same here, or a throw a not implemented error

File: packages/studio-base/src/players/RosbridgePlayer.ts [hunk 5]
```diff
@@ -645,25 +744,8 @@ export default class RosbridgePlayer implements Player {
...
-    for (const { topic, schemaName: datatype } of this.#advertisements) {
-      const roslibTopic = new roslib.Topic({
-        ros: this.#rosClient,
-        name: topic,
-        messageType: datatype,
-        queue_size: 0,
-      });
-      this.#topicPublishers.set(topic, roslibTopic);
-      roslibTopic.advertise();
-    }
```
Comment (charlielito) [inline packages/studio-base/src/players/RosbridgePlayer.ts]: same here, dont remove imple but comment and add a TODO

File: packages/studio-base/src/players/RosbridgePlayer.ts [hunk 6]
```diff
@@ -476,14 +511,133 @@ export default class RosbridgePlayer implements Player {
...
+            message.width = pointsFloat32.length/3;
+            message.height = 1;
+            message.point_step = 12;
+            message.row_step = message.width * message.point_step;
+            
+            message.data = points;
+          }
+
+          const innerMessage = message;
+          console.log ( innerMessage );
```
Comment (charlielito) [inline packages/studio-base/src/players/RosbridgePlayer.ts]: split this in different methods, for example, `parse_scan`, `parse_occupancy_grid` and so on so it is easier to read

File: packages/studio-base/src/players/data.json
Comment (charlielito) [inline packages/studio-base/src/players/data.json:1]: whaat? didn't understand your comment about this file, why is this needed???

Comment (charlielito) [reply packages/studio-base/src/players/data.json:1]: This has all our topics, I don't like it

**Review by charlielito — CHANGES_REQUESTED** [review]

File: packages/studio-base/src/players/RosbridgePlayer.ts (hunk 1 above)
Comment (charlielito) [reply packages/studio-base/src/players/RosbridgePlayer.ts]: dont remove old impl

**Comment by charlielito** [comment]
Discussing with @pepisg we think this is a great contribution to the ecosystem! But to be able to advertise it we cannot brake the rosbridge support. For that we wold need to add like an option to select the bridge implementation between rosboard and rosbridge. How hard do you think this would be?

File: packages/studio-base/src/players/RosbridgePlayer.ts [hunk 7]
```diff
@@ -476,14 +477,129 @@ export default class RosbridgePlayer implements Player {
...
+                    console.error('Error decoding image:', error);
+                });
+          }
+          else if ( message._topic_type === 'nav_msgs/msg/OccupancyGrid' ) {
+            let rdata = message._data_jpeg;
+
+            if ( this.#cachedGrid != undefined )
+                message.data = this.#cachedGrid;
+
+            // Decode the base64 JPEG to pixel data
```
Comment (charlielito) [inline packages/studio-base/src/players/RosbridgePlayer.ts]: fix comment

File: packages/studio-base/src/players/RosbridgePlayer.ts [hunk 8]
```diff
@@ -476,14 +477,129 @@ export default class RosbridgePlayer implements Player {
...
+                    // Assign the decoded RGB pixel data to message.data
+                    const decodedA = new Int8Array(pixelData); // Uint8Array
+
+                    const sumsArray = [];
+                    let alen = decodedA.length;
+                    if (decodedA != undefined) {
+                        for (let i=0; i < alen; i+=3) {
+                            sumsArray.push( 0.3 * (decodedA[i] || 0)
+										  + 0.59 * (decodedA[i+1] || 0)
+										  + 0.11 * (decodedA[i+2] || 0) );
```
Comment (charlielito) [inline packages/studio-base/src/players/RosbridgePlayer.ts]: what is this black magic? but a comment about those apparently random numbers

File: packages/studio-base/src/players/RosbridgePlayer.ts [hunk 9]
```diff
@@ -476,14 +477,129 @@ export default class RosbridgePlayer implements Player {
...
           const bytes = new Uint8Array(buffer);
-          const innerMessage = messageReader.readMessage(bytes);
+          // const innerMessage = messageReader.readMessage(bytes);
+          if ( message._topic_type === 'sensor_msgs/msg/LaserScan' ) {
+            let rbounds = message._ranges_uint16.bounds;
+
+            let rdata = this._base64decode(message._ranges_uint16.points);
+
+            let rview = new DataView(rdata);
+            let num_ranges = rdata.byteLength / 2;
```
Comment (charlielito) [inline packages/studio-base/src/players/RosbridgePlayer.ts]: why / 2 ?

File: packages/studio-base/src/players/RosbridgePlayer.ts [hunk 10]
```diff
@@ -476,14 +477,129 @@ export default class RosbridgePlayer implements Player {
...
+                .catch((error) => {
+                    console.error('Error decoding image:', error);
+                });
+          }
+          else if ( message._topic_type === 'sensor_msgs/msg/PointCloud2' ) {
+            let rdata = this._base64decode(message._data_uint16.points)
+            console.log ( rdata );
+            let rview = new DataView(rdata);
+
+            let num_ranges = rdata.byteLength / 6;
```
Comment (charlielito) [inline packages/studio-base/src/players/RosbridgePlayer.ts]: why / 6?

File: packages/studio-base/src/players/RosbridgePlayer.ts [hunk 11]
```diff
@@ -476,14 +477,129 @@ export default class RosbridgePlayer implements Player {
...
+
+            //const buffer = paddedPointsFloat32.buffer;
+            const buffer = pointsFloat32.buffer;
+            const points: Uint8Array = new Uint8Array(buffer);
+
+            message.width = pointsFloat32.length/3;
+            message.height = 1;
+            message.point_step = 12;
+            message.row_step = message.width * message.point_step;
+
```
Comment (charlielito) [inline packages/studio-base/src/players/RosbridgePlayer.ts]: please add comments about everything. I see a lot of constants that are hard to see why their values

File: packages/studio-base/src/players/RosbridgePlayer.ts [hunk 12]
```diff
@@ -476,14 +477,129 @@ export default class RosbridgePlayer implements Player {
...
+              pointsFloat32[i * 3 + 2] = this.scaleBackArray(z, zmin, zmax);
+            }
+
+            //const buffer = paddedPointsFloat32.buffer;
+            const buffer = pointsFloat32.buffer;
+            const points: Uint8Array = new Uint8Array(buffer);
+
+            message.width = pointsFloat32.length/3;
+            message.height = 1;
+            message.point_step = 12;
```
Comment (charlielito) [inline packages/studio-base/src/players/RosbridgePlayer.ts]: why is this hardcoded to those values?

File: packages/studio-base/src/players/RosbridgePlayer.ts [hunk 13]
```diff
@@ -476,14 +477,129 @@ export default class RosbridgePlayer implements Player {
...
+      this.#rosClient.addTopicCallback(topicName, (message) => {
         if (!this.#providerTopics) {
           return;
         }
         try {
           const buffer = (message as { bytes: ArrayBuffer }).bytes;
           const bytes = new Uint8Array(buffer);
-          const innerMessage = messageReader.readMessage(bytes);
+          // const innerMessage = messageReader.readMessage(bytes);
+          if ( message._topic_type === 'sensor_msgs/msg/LaserScan' ) {
```
Comment (pepisg) [inline packages/studio-base/src/players/RosbridgePlayer.ts]: this function got too long. Would it be possible to put the decoding for the different message types on independent functions so that we get a structure like this?

```
addTopicCallback
{
  if Laserscan
  {
    decode_laserscan()
  } 
  if PointCloud
  {
    decode_pointcloud()
  }
...
...
```

Instead of having all the decoding logic just placed in the function?

File: packages/studio-base/src/players/RosbridgePlayer.ts [hunk 14]
```diff
@@ -45,48 +45,34 @@ const log = Log.getLogger(__dirname);
...
-        if (!acc.has(value)) {
-          acc.set(value, new Set());
-        }
-        acc.get(value)?.add(node);
-      }
-    },
-    new Map<string, Set<string>>(),
-  );
-}
-
```
Comment (pepisg) [inline packages/studio-base/src/players/RosbridgePlayer.ts]: I'd prefer you commenting out the old parts instead of deleting them. In the future it would be easier to recover a functionality we may have removed for simplicity in this first integration

File: packages/studio-base/src/players/RosbridgePlayer.ts [hunk 15]
```diff
@@ -425,6 +415,11 @@ export default class RosbridgePlayer implements Player {
     this.#emitState();
   }
 
+  public scaleBackArray(value: number, min: number, max: number): number {
```
Comment (pepisg) [inline packages/studio-base/src/players/RosbridgePlayer.ts]: this isn't really taking in an array as the python version did. Could you rename to `scale_back_int16` or something similar?

File: packages/studio-base/src/players/RosbridgePlayer.ts [hunk 16]
```diff
@@ -546,85 +666,29 @@ export default class RosbridgePlayer implements Player {
...
+	  this.#getServiceType("");
+	  this.#refreshSystemState();
+	  return;
   }
 
   public setParameter(_key: string, _value: ParameterValue): void {
     throw new Error("Parameter editing is not supported by the Rosbridge connection");
   }
 
   public publish({ topic, msg }: PublishPayload): void {
```
Comment (pepisg) [inline packages/studio-base/src/players/RosbridgePlayer.ts]: can this function be implemented? if not, please leave a comment saying this is not yet implemented, however I think our rosboard transport does support sending messages from the client to the server

**Review by pepisg — CHANGES_REQUESTED** [review]
Some minor comments. Thanks for this very useful work

File: packages/studio-base/src/players/RosbridgePlayer.ts (hunk 16 above)
Comment (franyol) [reply packages/studio-base/src/players/RosbridgePlayer.ts]: We can expand our [custom rosboard client module](packages/studio-base/src/players/rosboardClient.ts) easily to achieve that.

**Comment by franyol** [comment]
Last commits notes:
* special types decoding functions where added to enhance code readability
* Code for unimplemented functionalities where bringed back inside commented blocks
* The code still needs the data.json file containing topic types and descriptions to work propperly, since the project parses datatypes which are used along other modules, we need to find a workaround to get this information, whether rosboard is able to bring it or not

File: packages/studio-base/src/players/RosbridgePlayer.ts (hunk 13 above)
Comment (charlielito) [reply packages/studio-base/src/players/RosbridgePlayer.ts]: troll haha ya le habia dicho eso

File: packages/studio-base/src/players/RosbridgePlayer.ts (hunk 14 above)
Comment (charlielito) [reply packages/studio-base/src/players/RosbridgePlayer.ts]: Same jaja

File: packages/studio-base/src/players/RosbridgePlayer.ts [hunk 17]
```diff
@@ -436,6 +448,137 @@ export default class RosbridgePlayer implements Player {
...
+    const binary_string = window.atob(base64);
+    const len = binary_string.length;
+    const bytes = new Uint8Array(len);
+    for (let i = 0; i < len; i++) {
+        bytes[i] = binary_string.charCodeAt(i);
+    }
+    return bytes.buffer;
+  }
+
+  public decode_LaserScan_msg (message: any): void {
```
Comment (charlielito) [inline packages/studio-base/src/players/RosbridgePlayer.ts]: For all these helper methods please use camelCase, since that is the convention in typescript methods

File: packages/studio-base/src/players/RosbridgePlayer.ts [hunk 18]
```diff
@@ -436,6 +448,137 @@ export default class RosbridgePlayer implements Player {
...
+	  });
+  }
+
+  public decode_OccupancyGrid_msg (message: any): void {
+	  let rdata = message._data_jpeg;
+
+	  if ( this.#cachedGrid != undefined )
+		  message.data = this.#cachedGrid;
+
+	  // Decode the base64 JPEG to pixel data
```
Comment (charlielito) [inline packages/studio-base/src/players/RosbridgePlayer.ts]: Fix comment!!

File: packages/studio-base/src/players/RosbridgePlayer.ts (hunk 8 above)
Comment (franyol) [reply packages/studio-base/src/players/RosbridgePlayer.ts]: I added a comment to clarify this: 
```typescript
          // Transform RGB to grayscale using luminocity method coefficients
          for (let i = 0; i < alen; i += 3) {
            sumsArray.push(
              0.3 * (decodedA[i] || 0) +
                0.59 * (decodedA[i + 1] || 0) +
                0.11 * (decodedA[i + 2] || 0),
            );
          }
```
So now it's clear those numbers are supposed to be coeficients to achieve RGB to grayscale convertion using luminocity method

File: packages/studio-base/src/players/RosbridgePlayer.ts (hunk 9 above)
Comment (franyol) [reply packages/studio-base/src/players/RosbridgePlayer.ts]: We are transforming an array with data of 16 bits into an array of float 32 bits, so `num_ranges` there is used as the length of the 32 bits data array.

File: packages/studio-base/src/players/RosbridgePlayer.ts (hunk 10 above)
Comment (franyol) [reply packages/studio-base/src/players/RosbridgePlayer.ts]: Similar as the application explained above, we are converting a uint16 array into a Float32 array, thus `rdata.byteLength/2` is the length of the transformed array, but in this case this array actually contains 3-dimentional data, thus, also dividing this length by 3 makes `num_ranges` useful to loop through the array as a 3-dimentional array with length=`num_ranges`

File: packages/studio-base/src/players/RosbridgePlayer.ts (hunk 16 above)
Comment (charlielito) [reply packages/studio-base/src/players/RosbridgePlayer.ts]: We are gonna left this as a future work

**Review by charlielito — APPROVED** [review]
